scripts/street_sign_recognizer.py

```diff
@@ -17,45 +19,139 @@ def __init__(self):
+
+        images = {
+        "left": '../images/leftturn_box_small.png',
+        "right": '../images/rightturn_box_small.png',
+        "uturn": '../images/uturn_box_small.png'
+        }
+
+        self.tm = TemplateMatcher(images)
+
+    def set_h_lb(self, val):
+        """ set hue lower bound """
+        self.hsv_lb[0] = val
+
+    def set_s_lb(self, val):
+        """ set saturation lower bound """
+        self.hsv_lb[1] = val
+
+    def set_v_lb(self, val):
+        """ set value lower bound """
+        self.hsv_lb[2] = val
+
+    def set_h_ub(self, val):
+        """ set hue upper bound """
+        self.hsv_ub[0] = val
+
+    def set_s_ub(self, val):
+        """ set saturation upper bound """
+        self.hsv_ub[1] = val
+
+    def set_v_ub(self, val):
+        """ set value upper bound """
+        self.hsv_ub[2] = val
 
     def process_image(self, msg):
         """ Process image messages from ROS and stash them in an attribute
             called cv_image for subsequent processing """
         self.cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
+        self.hsv_image = cv2.cvtColor(self.cv_image, cv2.COLOR_BGR2HSV)
+        self.threshold_image = cv2.inRange(self.hsv_image, self.hsv_lb, self.hsv_ub)
 
         left_top, right_bottom = self.sign_bounding_box()
+        #print left_top, right_bottom
         left, top = left_top
         right, bottom = right_bottom
 
+        # draw bounding box rectangle
+        cv2.rectangle(self.hsv_image, left_top, right_bottom, color=(255, 0, 0), thickness=5)
+
         # crop bounding box region of interest
         cropped_sign = self.cv_image[top:bottom, left:right]
+        grayscale_sign = cv2.cvtColor(cropped_sign, cv2.COLOR_BGR2GRAY)
+        print self.tm.predict(grayscale_sign)
 
-        # draw bounding box rectangle
-        cv2.rectangle(self.cv_image, left_top, right_bottom, color=(0, 0, 255), thickness=5)
+        
 
     def sign_bounding_box(self):
         """
         Returns
         -------
         (left_top, right_bottom) where left_top and right_bottom are tuples of (x_pixel, y_pixel)
             defining topleft and bottomright corners of the bounding box
+
+            I found this to be a good resource on contours
+            http://stackoverflow.com/questions/16538774/dealing-with-contours-and-bounding-rectangle-in-opencv-2-4-python-2-7
         """
-        # TODO: YOUR SOLUTION HERE
-        left_top = (200, 200)
-        right_bottom = (400, 400)
+        self.contour_image = copy.deepcopy(self.threshold_image)
+        contours, hierarchy = cv2.findContours(self.contour_image,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE)
+        # Find the index of the largest contour
+        areas = [cv2.contourArea(c) for c in contours]
+        max_index = np.argmax(areas)
+        cnt=contours[max_index]
+
+        x,y,w,h = cv2.boundingRect(cnt)
+        expand = 0.2
```

Comment:
A little comment like "The processing underestimates sign area, so we need to expand the bounding box" would have be nice to explain why you did the code that follows"